File: bindings/java-jni/src/main/java/org/maplibre/nativejni/geo/Feature.java
package org.maplibre.nativejni.geo;

import java.util.List;
import java.util.Objects;
import org.maplibre.nativejni.json.JsonValue;

/** Immutable GeoJSON feature descriptor. */
public record Feature(
    Geometry geometry, List<JsonValue.Member> properties, FeatureIdentifier identifier) {
  public Feature {
    Objects.requireNonNull(geometry, "geometry");
    properties = List.copyOf(properties);
    identifier = identifier == null ? FeatureIdentifier.nullValue() : identifier;
  }

  public Feature(Geometry geometry, List<JsonValue.Member> properties) {
    this(geometry, properties, FeatureIdentifier.nullValue());
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/geo/Feature.kt
package org.maplibre.nativeffi.geo

import org.maplibre.nativeffi.json.JsonValue

/** Immutable GeoJSON feature descriptor. */
public class Feature(
  public val geometry: Geometry,
  properties: List<JsonValue.Member>,
  public val identifier: FeatureIdentifier,
) {
  public val properties: List<JsonValue.Member> = properties.toList()

  override fun equals(other: Any?): Boolean =
    other is Feature &&
      geometry == other.geometry &&
      properties == other.properties &&
      identifier == other.identifier

  override fun hashCode(): Int {
    var result = geometry.hashCode()
    result = 31 * result + properties.hashCode()
    result = 31 * result + identifier.hashCode()
    return result
  }

  override fun toString(): String =
    "Feature(geometry=$geometry, properties=$properties, identifier=$identifier)"
}
